File: core/src/dc/targetman/level/MapLoader.java
package dc.targetman.level;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

import com.badlogic.gdx.maps.MapLayer;
import com.badlogic.gdx.maps.tiled.TiledMap;
import com.badlogic.gdx.maps.tiled.TiledMapTileLayer;
import com.badlogic.gdx.maps.tiled.objects.TiledMapTileMapObject;
import com.badlogic.gdx.math.Rectangle;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.math.Vector3;
import com.google.common.collect.ImmutableSet;

import dc.targetman.mechanics.Alliance;
import dclib.geometry.Point;
import dclib.geometry.PolygonUtils;
import dclib.graphics.ScreenHelper;

public final class MapLoader {

	private final TiledMap map;
	private final ScreenHelper screenHelper;
	private final EntityFactory entityFactory;

	public MapLoader(final TiledMap map, final ScreenHelper screenHelper, final EntityFactory entityFactory) {
		this.map = map;
		this.screenHelper = screenHelper;
		this.entityFactory = entityFactory;
	}

	public final void createObjects() {
		createStaticObjects();
		createActors();
	}

	public final void createStaticObjects() {
		Map<Point, List<Vector2>> pointToWallVertices = new HashMap<Point, List<Vector2>>();
		TiledMapTileLayer collisionLayer = MapUtils.getCollisionLayer(map);
		for (int x = 0; x < collisionLayer.getWidth(); x++) {
			for (int y = 0; y < collisionLayer.getHeight(); y++) {
				if (collisionLayer.getCell(x, y) != null) {
					addTileVertices(pointToWallVertices, x, y);
				}
			}
		}
		Set<List<Vector2>> wallsVertices = ImmutableSet.copyOf(pointToWallVertices.values());
		for (List<Vector2> wallVertices : wallsVertices) {
			entityFactory.createWall(wallVertices);
		}
	}

	private void addTileVertices(final Map<Point, List<Vector2>> pointToWallVertices, final int x, final int y) {
		float[] vertices = PolygonUtils.createRectangleVertices(new Rectangle(x, y, 1, 1));
		List<Vector2> newVertices = PolygonUtils.toVectors(vertices);
		List<Vector2> wallVertices = newVertices;
		Point point = new Point(x, y);
		// due to the order at which tiles are iterated, current neighbors can only exist at left or bottom
		Point left = new Point(x - 1, y);
		Point bottom = new Point(x, y - 1);
		if (pointToWallVertices.containsKey(left)) {
			wallVertices = pointToWallVertices.get(left);
			PolygonOperations.union(wallVertices, newVertices);
		} else if (pointToWallVertices.containsKey(bottom)) {
			wallVertices = pointToWallVertices.get(bottom);
			PolygonOperations.union(wallVertices, newVertices);
		}
		pointToWallVertices.put(point, wallVertices);
	}

	private void createActors() {
		MapLayer layer = map.getLayers().get(1);
		for (TiledMapTileMapObject object : layer.getObjects().getByType(TiledMapTileMapObject.class)) {
			Vector3 position = getWorldPosition(object, 0);
			String allianceString = object.getTile().getProperties().get(Alliance.class.getSimpleName(), String.class);
			entityFactory.createStickman(position, Alliance.valueOf(allianceString));
		}
	}

	private Vector3 getWorldPosition(final TiledMapTileMapObject object, final float z) {
		Vector2 worldPosition = screenHelper.toWorldUnits(object.getX(), object.getY());
		return new Vector3(worldPosition.x, worldPosition.y, 0);
	}

}

File: core/src/main/dc/targetman/level/MapLoader.kt
package dc.targetman.level

import com.badlogic.gdx.graphics.g2d.TextureRegion
import com.badlogic.gdx.maps.tiled.TiledMap
import com.badlogic.gdx.maps.tiled.TiledMapTileLayer
import com.badlogic.gdx.maps.tiled.objects.TiledMapTileMapObject
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.math.Vector3
import dc.targetman.character.CharacterFactory
import dc.targetman.mechanics.Alliance
import dc.targetman.physics.Interactivity
import dc.targetman.physics.collision.CollisionCategory
import dc.targetman.physics.collision.Material
import dclib.epf.Entity
import dclib.epf.parts.TransformPart
import dclib.geometry.PolygonOperations
import dclib.geometry.PolygonUtils
import dclib.geometry.toVector3
import dclib.graphics.TextureUtils
import dclib.map.Cell
import dclib.map.MapUtils
import dclib.physics.Box2dTransform
import dclib.physics.Box2dUtils

class MapLoader(private val map: TiledMap, private val factoryTools: FactoryTools) {
    private val scale = 1 / MapUtils.getPixelsPerUnit(map)
    private val characterFactory = CharacterFactory(factoryTools)
    private val hullCache = mutableMapOf<TextureRegion, FloatArray>()

    fun createObjects() {
        createWalls()
        createActors()
        createLadders()
    }

    fun createWalls() {
        val layer = map.layers[MapUtils.FOREGROUND_INDEX] as TiledMapTileLayer
        val tilesVertices = mutableListOf<List<Vector2>>()
        for (cell in MapUtils.getCells(layer)) {
            tilesVertices.add(createTileVertices(cell))
        }
        val unionTileVertices = PolygonOperations.union(tilesVertices)
        for (tileVertices in unionTileVertices) {
            createWall(tileVertices)
        }
    }

    fun createCharacter(characterPath: String, position: Vector3, alliance: Alliance): Entity {
        return characterFactory.create(characterPath, 2f, position, alliance)
    }

    private fun createTileVertices(cell: Cell): List<Vector2> {
        val tileCell = cell.cell
        var convexHull = getHull(tileCell.tile.textureRegion)
        PolygonUtils.scale(convexHull, scale)
        if (tileCell.flipHorizontally) {
            convexHull = PolygonUtils.flipX(convexHull)
        }
        if (tileCell.flipVertically) {
            convexHull = PolygonUtils.flipY(convexHull)
        }
        convexHull = PolygonUtils.shift(convexHull, cell.point.x.toFloat(), cell.point.y.toFloat())
        return PolygonUtils.toVectors(convexHull)
    }

    private fun createActors() {
        val layer = map.layers.get(1)
        for (mapObject in layer.objects.getByType(TiledMapTileMapObject::class.java)) {
            val position = Vector2(mapObject.x, mapObject.y).scl(scale).toVector3()
            val allianceString = mapObject.tile.properties.get(Alliance::class.java.simpleName, String::class.java)
            val alliance = Alliance.valueOf(allianceString)
            val characterPath = if (alliance == Alliance.PLAYER) "characters/cyborg.json" else "characters/dummy.json"
            createCharacter(characterPath, position, alliance)
        }
    }

    private fun createWall(vertices: List<Vector2>) {
        val body = Box2dUtils.createStaticBody(factoryTools.world, PolygonUtils.toFloats(vertices))
        Box2dUtils.setFilter(body, CollisionCategory.STATIC, CollisionCategory.ALL)
        val transform = Box2dTransform(body)
        val entity = Entity(TransformPart(transform))
        entity.addAttributes(Material.METAL)
        factoryTools.entityManager.add(entity)
    }

    private fun createLadders() {
        val layer = map.layers[MapUtils.BACKGROUND_INDEX] as TiledMapTileLayer
        val tilesVertices = mutableListOf<List<Vector2>>()
        for (cell in MapUtils.getCells(layer)) {
            if (cell.cell.tile.properties.containsKey(Interactivity.CLIMB.toString())) {
                tilesVertices.add(createTileVertices(cell))
            }
        }
        val unionTileVertices = PolygonOperations.union(tilesVertices)
        for (tileVertices in unionTileVertices) {
            createLadder(tileVertices)
        }
    }

    private fun createLadder(vertices: List<Vector2>) {
        val body = Box2dUtils.createStaticBody(factoryTools.world, PolygonUtils.toFloats(vertices))
        Box2dUtils.setFilter(body, null, CollisionCategory.BOUNDS)
        Box2dUtils.setSensor(body, true)
        val transform = Box2dTransform(body)
        val entity = Entity(TransformPart(transform))
        entity.addAttributes(Interactivity.CLIMB)
        factoryTools.entityManager.add(entity)
    }

    private fun getHull(region: TextureRegion): FloatArray {
        return hullCache.getOrPut(region, { TextureUtils.createConvexHull(region) }).copyOf()
    }
}
